### ISSUE TITLE
Stricter `no-undef` default

### ISSUE THREAD

**Issue by BridgeAR** [eslint/eslint#10203]
**Tell us about your environment**

* **ESLint Version:** 5.0.0-alpha
* **Node Version:** 10.0.0-pre
* **npm Version:** 5.6

**What parser (default, Babel-ESLint, etc.) are you using?**

Default

**Please show your full configuration:**

`no-undef: error`

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
if (typeof undef === 'string') {
  throw new Error('foobar')
}
```

**What did you expect to happen?**

An error should be reported. This is possible by using the `typeof` option. I personally feel this option should be activated by default as I do not see a reason why it such statements are not validated by default.

**What actually happened? Please include the actual, raw output from ESLint.**

No error is reported by default.

**Comment by platinumazure** [comment]
Hi @BridgeAR, thanks for the issue.

As far as I can tell, you're aware of the rule option for ensuring typeof lines are also checked. So the request here seems to be that you want that option on by default. That's a pretty hefty request since we need to make such a change on a semver-major release and users might need to upgrade their configurations.

Is there really a strong case for making this change? Or, is there some other change we could make (e.g., in documentation) to make this option more discoverable?

**Comment by BridgeAR** [comment]
Yes, I am aware of the option since I ran into an [issue](https://github.com/nodejs/node/pull/19926) in Node.js. I was super confused when I realized that variables in combination with `typeof` were not checked and I thought it would be a bug at first glance. 

I guess most people switched a while ago from jshint to eslint and thought it would work identical in this case. I understand that this is a breaking change but at the very same time I actually believe it is a bug fix. So I personally believe it is best to change this in 5.0.0. I am not sure why this was hidden behind a option in the first place.

**Comment by mysticatea** [comment]
Context: #3684

The current `no-undef` behavior is according to the language specification. We can use undefined variables as the operand of `typeof` operator in JavaScript. However, "valid in the language spec" and "valid in an application logic" are different, so I added the option at that time.

I'm OK to change the default behavior as a breaking change -- this is not a bug because it's an intended behavior.

**Comment by platinumazure** [comment]
I'm mildly opposed to making a breaking change here because of the fact that this is behaving like the language spec, and we do have the option available. I'm not convinced that there is much benefit to making the breaking change.

**Comment by BridgeAR** [comment]
What about a small poll? I guess most people actually do not know about the option and I also guess a lot of people would rather have the "strict" version as default. It would also be interesting to know how much code would actually break with this change but that is difficult to measure.

**Comment by ilyavolodin** [comment]
This is one of those situations where both approaches are valid, and we will never get consensus from the users as to what should be the default behavior. If we were to make the breaking change, I'm pretty sure we will get issues from users to revert it back to the current behavior. As we have an option to control this behavior, I think it's reasonable to leave it up to the user to decide what they want. So, in short, I'm mildly opposed to making this breaking change, because it will negatively affect existing users who are aware of the current behavior and rely on it.

**Comment by platinumazure** [comment]
@eslint/eslint-team Anyone want to champion this (and try to work through the disagreement)? If we don't get a champion on this proposal, I think we need to close this issue.

**Comment by nzakas** [comment]
Apologies, I close this by accident. Reopened.

**Comment by not-an-aardvark** [comment]
In [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/3e40976a2ebb9481405d4c3a358bc3191ecd234a/notes/2019/2019-02-28.md#stricter-no-undef-default), the TSC decided to table this issue for v6.0.0 but possibly reconsider it for v7.0.0.

**Comment by btmills** [comment]
Since there's not a clear direction here, we decided in today's TSC meeting to move this to the v8.0.0 project as a possible change for that release. @mysticatea, are you still interested in championing this?

**Comment by btmills** [comment]
Thanks all for the discussion on this. Since we have split opinions on this issue and there hasn't been any further discussion in almost 3 years, I'll close this instead of bumping it again to another release.
